test_src/sljitTest.c

```diff
@@ -11529,245 +11538,205 @@ static void test92(void)
 	executable_code code;
 	struct sljit_compiler *compiler = sljit_create_compiler(NULL, NULL);
 	struct sljit_label *label;
-	struct sljit_jump *jump;
-	sljit_sw buf[32];
-	sljit_s32 i;
+	struct sljit_jump *jump, *skip;
+	sljit_sw buf[12];
 
 	if (verbose)
 		printf("Run test92\n");
 
 	FAILED(!compiler, "cannot create compiler\n");
 
-	for (i = 0; i < 32; i++)
-		buf[i] = -1;
-
 	buf[0] = -4678;
-	*(sljit_u8*)(buf + 2) = 178;
-	*(sljit_u8*)(buf + 5) = 211;
-	*(sljit_u16*)(buf + 9) = 17897;
-	*(sljit_u16*)(buf + 12) = 57812;
-	*(sljit_u32*)(buf + 15) = 1234567890;
-	*(sljit_u32*)(buf + 17) = 987609876;
-	*(sljit_u8*)(buf + 20) = 192;
-	buf[23] = 6359;
-	((sljit_u8*)(buf + 26))[1] = 105;
-	((sljit_u8*)(buf + 28))[2] = 13;
-	((sljit_u16*)(buf + 30))[1] = 14876;
+	buf[1] = HV(sljit_u8, 0, 178);
+	buf[2] = HV(sljit_u8, 0, 211);
+	buf[3] = HV(sljit_u16, 0, 17897);
+	buf[4] = HV(sljit_u16, 0, 57812);
+	buf[5] = HV(sljit_u32, 0, 1234567890);
+	buf[6] = HV(sljit_u32, 0, 987609876);
+	buf[7] = HV(sljit_u8, 0,192);
+	buf[8] = 6359;
+	buf[9] = HV(sljit_u8, 1, 105);
+	buf[10] = HV(sljit_u8, 2, 13);
+	buf[11] = HV(sljit_u16, 1, 14876);
 
 	sljit_emit_enter(compiler, 0, SLJIT_ARGS1(VOID, P), 5, 5, 0, 0, 2 * sizeof(sljit_sw));
 
 	label = sljit_emit_label(compiler);
 	sljit_emit_atomic_load(compiler, SLJIT_MOV, SLJIT_R1, SLJIT_S0);
-	/* buf[1] */
-	sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_MEM1(SLJIT_S0), sizeof(sljit_sw), SLJIT_R1, 0);
-	sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_R0, 0, SLJIT_R1, 0);
+	jump = sljit_emit_cmp(compiler, SLJIT_NOT_EQUAL, SLJIT_R1, 0, SLJIT_MEM1(SLJIT_S0), 0);
```

Comment:
I would keep the original logic of storing the result of loads.

Reply:
Wouldn't this require then memory barriers to be added?, moving the store outside of the critical loop seems to work, but without a memory barrier it might get reordered and would cause failures (although not as obvious, as the infinite loop)

When refactoring, I thought about using a second array for the stores, but with a possible granule of 512 words, it will need ugly tricks to keep it safe anyway.

Reply:
On ARM the doc says in "Load-Acquire, Store-Release" section:
- A Load-Acquire places no additional ordering constraints on any loads or stores appearing before the
Load-Acquire.
- The Store-Release places no additional ordering constraints on any loads or stores appearing after the
Store-Release instruction.

If barriers needs to be added on some cpus, the instruction should implicitly add them.

Reply:
Note that (even before LSE) in aarch64 there are 2 pairs of instructions for its LL/SC.

Our implementation uses the weak (and better performing) pair of `ldxr`/`stxr` but their semantics don't correspond to what is used on other processors and that seem to match better the strong (and slower) pair (`ldaxr`/`stlxr`).

Not that anyone should, but if someone would implement a futex or mutex with sljit, it might work fine in x86/s390x and break with ARM (and probably the other RISC CPUs if implemented using LL/SC).

I have to admit that without a clear understanding on how this API is to be used, I am not sure if it is a problem but it might be something worth considering by maybe adding a "weak" parameter that would be a NOOP in strong ordered architectures and allow selecting the right pair in weak ordered ones like ARM.  

If this is to be modeled like C11/C++11 atomics then maybe a full set of memory_order options might be needed as well.